components/web_discovery/browser/ecdh_aes.cc

```diff
@@ -0,0 +1,137 @@
+
+#include <array>
+
+#include "base/base64.h"
+#include "base/logging.h"
+#include "base/ranges/algorithm.h"
+#include "base/strings/string_number_conversions.h"
+#include "crypto/random.h"
+#include "crypto/sha2.h"
+#include "third_party/boringssl/src/include/openssl/aead.h"
+#include "third_party/boringssl/src/include/openssl/ec.h"
+#include "third_party/boringssl/src/include/openssl/ec_key.h"
+#include "third_party/boringssl/src/include/openssl/ecdh.h"
+#include "third_party/boringssl/src/include/openssl/nid.h"
+
+namespace web_discovery {
+
+namespace {
+
+constexpr size_t kAesKeySize = 16;
+constexpr size_t kAesTagLength = 16;
+constexpr size_t kIvSize = 12;
+constexpr size_t kKeyMaterialSize = 32;
+// P-256 field size * 2 + type byte
+constexpr size_t kComponentOctSize = 32 * 2 + 1;
+// type byte + public component + initialization vector
+constexpr size_t kEncodedPubKeyAndIv = 1 + kComponentOctSize + kIvSize;
+constexpr uint8_t kP256TypeByte = 0xea;
+
+bssl::UniquePtr<EC_KEY> CreateECKey() {
+  return bssl::UniquePtr<EC_KEY>(
+      EC_KEY_new_by_curve_name(NID_X9_62_prime256v1));
+}
+
+}  // namespace
+
+AESEncryptResult::AESEncryptResult(std::vector<uint8_t> data,
+                                   std::string encoded_public_component_and_iv)
+    : data(data),
+      encoded_public_component_and_iv(encoded_public_component_and_iv) {}
+
+AESEncryptResult::~AESEncryptResult() = default;
+AESEncryptResult::AESEncryptResult(const AESEncryptResult&) = default;
+
+std::optional<AESEncryptResult> DeriveAESKeyAndEncrypt(
+    const std::vector<uint8_t>& server_pub_key,
+    const std::vector<uint8_t>& data) {
+  auto client_private_key = CreateECKey();
+
+  if (!client_private_key) {
+    VLOG(1) << "Failed to init P-256 curve";
+    return std::nullopt;
+  }
+
+  bssl::UniquePtr<EC_POINT> server_public_point(EC_POINT_new(EC_group_p256()));
+  if (!server_public_point) {
+    VLOG(1) << "Failed to init EC public point";
+    return std::nullopt;
+  }
+
+  if (!EC_POINT_oct2point(EC_group_p256(), server_public_point.get(),
+                          server_pub_key.data(), server_pub_key.size(),
+                          nullptr)) {
+    VLOG(1) << "Failed to load server public key data into EC point";
+    return std::nullopt;
+  }
+
+  if (!EC_KEY_generate_key(client_private_key.get())) {
+    VLOG(1) << "Failed to generate client EC key";
+    return std::nullopt;
+  }
+
+  uint8_t shared_key_material[kKeyMaterialSize];
+  if (!ECDH_compute_key(shared_key_material, kKeyMaterialSize,
+                        server_public_point.get(), client_private_key.get(),
+                        nullptr)) {
+    VLOG(1) << "Failed to set derive key via ECDH";
+    return std::nullopt;
+  }
+
+  auto key_material_hash = crypto::SHA256Hash(shared_key_material);
+
+  auto aes_key = std::vector<uint8_t>(key_material_hash.begin(),
+                                      key_material_hash.begin() + kAesKeySize);
```

Comment:
Unless I'm wrong on what I'm reading we should be just using a span, rather than allocating a vector here:

```
auto aes_key = crypto::SHA256Hash(shared_key_material).first<kAesKeySize>();
```